connman/vpn/vpn-provider.c

```diff
@@ -1613,21 +1613,55 @@ static void connect_cb(struct vpn_provider *provider, void *user_data,
 						VPN_PROVIDER_ERROR_UNKNOWN);
 			break;
 		case ECANCELED:
-			/* fall through */
+			/*
+			 * ECANCELED means that agent dialog was canceled and
+			 * if the VPN has been connected make it disconnected.
+			 */
+			if (is_connected_state(provider->state)) {
+				if (provider->driver->set_state)
```

Comment:
In theory yes, in practice no. This `connect_cb()` is called only from a VPN plugin, which is not possible to run without the `driver`. So in practice that does not happen. In other places it is checked because the calls can come at times when driver may be uninitialized, or needs to be checked for.